### PR TITLE
Fix beacon-chain/sync Test Race

### PR DESCRIPTION
**Pull request by fgimenez**
Towards #377

Running the tests with `--features=race` I was getting:
```
WARNING: DATA RACE
Write at 0x00c42011d5a8 by goroutine 35:
  github.com/prysmaticlabs/prysm/beacon-chain/sync.TestSavingBlocksInSync()
      beacon-chain/sync/service_test.go:832 +0xb45
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Previous read at 0x00c42011d5a8 by goroutine 36:
  github.com/prysmaticlabs/prysm/beacon-chain/sync.(*Service).initialSync()
      bazel-out/k8-fastbuild/bin/proto/beacon/p2p/v1/linux_amd64_race_stripped/v1_go_proto~/github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1/messages.pb.go:223 +0x946
  github.com/prysmaticlabs/prysm/beacon-chain/sync.TestSavingBlocksInSync.func1()
      beacon-chain/sync/service_test.go:759 +0x74

Goroutine 35 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
  testing.runTests()
      GOROOT/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:978 +0x2cd
  main.main()
      bazel-out/k8-fastbuild/bin/beacon-chain/sync/linux_amd64_race_stripped/go_default_test~/testmain.go:97 +0x28e

Goroutine 36 (running) created at:
  github.com/prysmaticlabs/prysm/beacon-chain/sync.TestSavingBlocksInSync()
      beacon-chain/sync/service_test.go:758 +0x212
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
```
There was a problem with reusing `msg1` in the test and changing members while the initialization goroutine was still active. It is fixed by generating different block response messages for each assertion.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Thanks for this! Does this solve all of the race condition issues or just one?

If it solves all, can you add `--features=race` to the travis test?

**Comment by fgimenez** [comment]
> Does this solve all of the race condition issues or just one?

@prestonvanloon just one, there are currently 4 race conditions on master:
```
INFO: 104 processes: 104 linux-sandbox.
//beacon-chain/blockchain:go_default_test                       (cached) PASSED in 1.2s
//beacon-chain/powchain:go_default_test                         (cached) PASSED in 1.2s
//beacon-chain/simulator:go_default_test                        (cached) PASSED in 1.2s
//beacon-chain/utils:go_default_test                            (cached) PASSED in 1.1s
//client/attester:go_default_test                               (cached) PASSED in 1.5s
//client/contracts:go_default_test                              (cached) PASSED in 296.9s
//client/mainchain:go_default_test                              (cached) PASSED in 3.2s
//client/observer:go_default_test                               (cached) PASSED in 1.8s
//client/params:go_default_test                                 (cached) PASSED in 1.1s
//client/txpool:go_default_test                                 (cached) PASSED in 1.1s
//client/types:go_default_test                                  (cached) PASSED in 1.2s
//contracts:go_default_test                                     (cached) PASSED in 1.4s
//shared:go_default_test                                        (cached) PASSED in 18.1s
//shared/cmd:go_default_test                                    (cached) PASSED in 1.2s
//shared/database:go_default_test                               (cached) PASSED in 1.4s
//shared/p2p:go_feed_concurrent_write_test                      (cached) PASSED in 1.7s
//beacon-chain/node:go_default_test                                      PASSED in 1.9s
//beacon-chain/rpc:go_default_test                                       PASSED in 1.1s
//client/node:go_default_test                                            PASSED in 1.8s
//client/rpcclient:go_default_test                                       PASSED in 1.1s
//beacon-chain/sync:go_default_test                                      FAILED in 5 out of 5 in 0.2s
  Stats over 5 runs: max = 0.2s, min = 0.1s, avg = 0.2s, dev = 0.0s
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/beacon-chain/sync/go_default_test/test_attempts/attempt_1.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/beacon-chain/sync/go_default_test/test_attempts/attempt_2.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/beacon-chain/sync/go_default_test/test_attempts/attempt_3.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/beacon-chain/sync/go_default_test/test_attempts/attempt_4.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/beacon-chain/sync/go_default_test/test.log
//client/proposer:go_default_test                                        FAILED in 5 out of 5 in 3.8s
  Stats over 5 runs: max = 3.8s, min = 3.7s, avg = 3.8s, dev = 0.1s
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/proposer/go_default_test/test_attempts/attempt_1.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/proposer/go_default_test/test_attempts/attempt_2.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/proposer/go_default_test/test_attempts/attempt_3.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/proposer/go_default_test/test_attempts/attempt_4.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/proposer/go_default_test/test.log
//client/syncer:go_default_test                                          FAILED in 5 out of 5 in 0.9s
  Stats over 5 runs: max = 0.9s, min = 0.9s, avg = 0.9s, dev = 0.0s
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/syncer/go_default_test/test_attempts/attempt_1.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/syncer/go_default_test/test_attempts/attempt_2.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/syncer/go_default_test/test_attempts/attempt_3.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/syncer/go_default_test/test_attempts/attempt_4.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/syncer/go_default_test/test.log
//shared/p2p:go_default_test                                             FAILED in 5 out of 5 in 1.7s
  Stats over 5 runs: max = 1.7s, min = 1.3s, avg = 1.5s, dev = 0.2s
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/shared/p2p/go_default_test/test_attempts/attempt_1.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/shared/p2p/go_default_test/test_attempts/attempt_2.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/shared/p2p/go_default_test/test_attempts/attempt_3.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/shared/p2p/go_default_test/test_attempts/attempt_4.log
  /home/fgimenez/.cache/bazel/_bazel_fgimenez/512843e3dd6bfc3af7fbee03058a93ee/execroot/__main__/bazel-out/k8-fastbuild/testlogs/shared/p2p/go_default_test/test.log
```
This PR solves `//beacon-chain/sync:go_default_test`. I think we can solve `//client/syncer:go_default_test ` and part of `//client/proposer:go_default_test ` too, will propose PRs for them next.

The rest of the issues come from the `libp2p-go` thirdparty library. The problem is tracked in this upstream issue https://github.com/libp2p/go-libp2p/issues/257, they are aiming to switch the mdns library but seem to be blocked by https://github.com/grandcat/zeroconf/issues/38

While this is not solved we could add the `--features=race` to travis command as you suggest and change the bazel config for the tests that are known to be racy, similar to what we did here https://github.com/prysmaticlabs/prysm/pull/321/files#diff-f127b9ba81d65ec476bc4edb33289780 but the other way around (setting `race=off` for the racy tests). This way the race detector would be enabled by default, WDYT?

**Comment by prestonvanloon** [comment]
OK, this is great progress nonetheless.

I recommend adding a race testing exclusion on those targets for now:

```python
    race = "off", # TODO(#377): fix issues with race detection testing.
```

And adding `--features=race` to the travis yaml. Maybe this will prevent future issues while we work through the existing ones.
Thanks!

File: beacon-chain/sync/service_test.go [hunk 1]
```diff
@@ -785,22 +785,26 @@ func TestSavingBlocksInSync(t *testing.T) {
 		t.Fatalf("unable to get hash of crystallized state: %v", err)
 	}
 
-	block := &pb.BeaconBlock{
-		MainChainRef:          []byte{1, 2, 3},
-		ParentHash:            generichash,
-		SlotNumber:            uint64(20),
-		CrystallizedStateHash: crystallizedStateHash[:],
-	}
+	getBlockResponseMsg := func(slotNumber uint64) p2p.Message {
```
Comment (terenc3t) [inline beacon-chain/sync/service_test.go]: This is neat!

**Comment by fgimenez** [comment]
just pushed changes to disable race detector on `client/contracts`, was timing out with it.

**Comment by fgimenez** [comment]
@prestonvanloon great suggestion, done PTAL.

**Comment by fgimenez** [comment]
I have ready changes for preventing the `client/syncer` race, should I push them here or in a separate PR once this one is merged? The changes in this one for enabling the race detector by default would be required to verify that the fix works.

**Comment by prestonvanloon** [comment]
@fgimenez In another PR please. I'll merge this now

**Comment by fgimenez** [comment]
With the removal of `client/syncer` in #395 the only remaining race is in `shared/p2p` caused by `libp2p-go`. I'll keep an eye on https://github.com/libp2p/go-libp2p/issues/257 and https://github.com/grandcat/zeroconf/issues/38.

**Pull request merged**